File: javascript/functions/queryShipment.js
/**
 * queryShipment - Gönderi Sorgulama
 * Müşterinin kendi referans kodu ile gönderi geçmişini raporlayan servis.
 */

import { sendSoapRequest, parseXml, escapeXml } from '../YurticiKargoClient.js';

/**
 * @param {Object} config - {username, password, language, endpoint}
 * @param {Array<string>} keys - Sorgulanacak anahtarlar
 * @param {number} keyType - 0=CargoKey, 1=InvoiceKey
 * @param {boolean} addHistoricalData - Hareket geçmişi dahil mi
 * @param {boolean} onlyTracking - Sadece takip URL'si mi
 * @returns {Promise<Object>}
 */
export async function queryShipment(config, keys, keyType = 0, addHistoricalData = false, onlyTracking = false) {
  if (!Array.isArray(keys) || keys.length === 0) {
    throw new Error('keys parametresi boş olamaz');
  }

  const keysXml = keys
    .map(key => `      <keys>${escapeXml(String(key))}</keys>`)
    .join('\n');

  const soapBody = `
    <ship:queryShipment>
      <wsUserName>${escapeXml(config.username)}</wsUserName>
      <wsPassword>${escapeXml(config.password)}</wsPassword>
      <wsLanguage>${escapeXml(config.language)}</wsLanguage>
${keysXml}
      <keyType>${keyType}</keyType>
      <addHistoricalData>${addHistoricalData}</addHistoricalData>
      <onlyTracking>${onlyTracking}</onlyTracking>
    </ship:queryShipment>`;

  const responseXml = await sendSoapRequest(config.endpoint, soapBody);
  return parseQueryShipmentResponse(responseXml);
}

/**
 * queryShipment yanıtını parse eder
 */
function parseQueryShipmentResponse(xml) {
  const parsed = parseXml(xml);

  const outFlag = parseInt(extractValue(parsed, 'outFlag') || '-1');
  const outResult = extractValue(parsed, 'outResult') || '';
  const senderCustId = extractValue(parsed, 'senderCustId') || '';
  const count = parseInt(extractValue(parsed, 'count') || '0');

  const details = [];
  const queryDetails = extractAllElements(parsed, 'shippingDeliveryDetailVO');

  for (const detail of queryDetails) {
    const item = {
      cargoKey: extractValue(detail, 'cargoKey') || '',
      invoiceKey: extractValue(detail, 'invoiceKey') || '',
      operationStatus: extractValue(detail, 'operationStatus') || '',
      operationMessage: extractValue(detail, 'operationMessage') || '',
      errCode: parseInt(extractValue(detail, 'errCode') || '0'),
      errMessage: extractValue(detail, 'errMessage') || '',
      itemDetail: null
    };

    // itemDetail parse - ShippingDeliveryItemDetailVO
    const itemDetailNode = findElement(detail, 'shippingDeliveryItemDetailVO');
    if (itemDetailNode) {
      item.itemDetail = {
        trackingUrl: extractValue(itemDetailNode, 'trackingUrl') || '',
        receiverCustName: extractValue(itemDetailNode, 'receiverCustName') || '',
        departureUnitName: extractValue(itemDetailNode, 'departureUnitName') || '',
        deliveryDate: extractValue(itemDetailNode, 'deliveryDate') || '',
        deliveryTime: extractValue(itemDetailNode, 'deliveryTime') || '',
        cargoHistory: []
      };

      // Kargo geçmişi - cargoEventHistory
      const historyItems = extractAllElements(itemDetailNode, 'cargoEventHistory');
      for (const hist of historyItems) {
        item.itemDetail.cargoHistory.push({
          unitName: extractValue(hist, 'unitName') || '',
          eventName: extractValue(hist, 'eventName') || '',
          reasonName: extractValue(hist, 'reasonName') || '',
          eventDate: extractValue(hist, 'eventDate') || '',
          eventTime: extractValue(hist, 'eventTime') || ''
        });
      }
    }

    item.isSuccess = function() { return this.errCode === 0; };
    details.push(item);
  }

  return {
    outFlag,
    outResult,
    senderCustId,
    count,
    details,
    isSuccess() { return this.outFlag === 0; }
  };
}

function extractValue(obj, key) {
  if (!obj) return null;
  if (typeof obj === 'string') return null;

  if (obj[key] !== undefined) {
    if (typeof obj[key] === 'object' && obj[key] !== null) {
      return obj[key].__text || '';
    }
    return obj[key];
  }

  for (const k of Object.keys(obj)) {
    if (typeof obj[k] === 'object' && obj[k] !== null) {
      const found = extractValue(obj[k], key);
      if (found !== null) return found;
    }
  }
  return null;
}

function findElement(obj, elementName) {
  if (!obj || typeof obj !== 'object') return null;
  if (obj[elementName]) return obj[elementName];
  for (const k of Object.keys(obj)) {
    if (typeof obj[k] === 'object' && obj[k] !== null) {
      const found = findElement(obj[k], elementName);
      if (found) return found;
    }
  }
  return null;
}

function extractAllElements(obj, elementName) {
  const results = [];
  if (!obj || typeof obj !== 'object') return results;

  if (obj[elementName]) {
    const items = Array.isArray(obj[elementName]) ? obj[elementName] : [obj[elementName]];
    results.push(...items);
  }

  for (const k of Object.keys(obj)) {
    if (k !== elementName && typeof obj[k] === 'object' && obj[k] !== null) {
      results.push(...extractAllElements(obj[k], elementName));
    }
  }

  return results;
}

export default queryShipment;

File: go/functions/query_shipment.go
package functions

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/yurticikargo/api-client/models"
)

// buildQueryShipmentXML builds the SOAP XML for queryShipment (deliveryReportShipment).
func buildQueryShipmentXML(username, password, language string, keys []string, keyType int, addHistoricalData bool, onlyTracking bool) string {
	var sb strings.Builder

	sb.WriteString(`<?xml version="1.0" encoding="UTF-8"?>`)
	sb.WriteString(`<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ship="http://yurticikargo.com.tr/ShippingOrderDispatcherServices">`)
	sb.WriteString(`<soapenv:Header/>`)
	sb.WriteString(`<soapenv:Body>`)
	sb.WriteString(`<ship:deliveryReportShipment>`)
	sb.WriteString(fmt.Sprintf(`<wsUserName>%s</wsUserName>`, xmlEscape(username)))
	sb.WriteString(fmt.Sprintf(`<wsPassword>%s</wsPassword>`, xmlEscape(password)))
	sb.WriteString(fmt.Sprintf(`<wsLanguage>%s</wsLanguage>`, xmlEscape(language)))
	sb.WriteString(fmt.Sprintf(`<keys>%s</keys>`, xmlEscape(strings.Join(keys, ","))))
	sb.WriteString(fmt.Sprintf(`<keyType>%d</keyType>`, keyType))
	sb.WriteString(fmt.Sprintf(`<addHistoricalData>%s</addHistoricalData>`, boolToString(addHistoricalData)))
	sb.WriteString(fmt.Sprintf(`<onlyTracking>%s</onlyTracking>`, boolToString(onlyTracking)))
	sb.WriteString(`</ship:deliveryReportShipment>`)
	sb.WriteString(`</soapenv:Body>`)
	sb.WriteString(`</soapenv:Envelope>`)

	return sb.String()
}

// queryResponseEnvelope is used for XML parsing of the SOAP response.
type queryResponseEnvelope struct {
	XMLName xml.Name        `xml:"Envelope"`
	Body    queryResponseBody `xml:"Body"`
}

type queryResponseBody struct {
	QueryResponse deliveryReportResponse `xml:"deliveryReportShipmentResponse"`
}

type deliveryReportResponse struct {
	QueryResult queryResultXML `xml:"ShippingDeliveryResultVO"`
}

type queryResultXML struct {
	OutFlag      int              `xml:"outFlag"`
	OutResult    string           `xml:"outResult"`
	SenderCustId int              `xml:"senderCustId"`
	Count        int              `xml:"count"`
	Details      []queryDetailXML `xml:"shippingDeliveryDetailVO"`
}

type queryDetailXML struct {
	CargoKey         string              `xml:"cargoKey"`
	InvoiceKey       string              `xml:"invoiceKey"`
	OperationStatus  string              `xml:"operationStatus"`
	OperationMessage string              `xml:"operationMessage"`
	ErrCode          int                 `xml:"errCode"`
	ErrMessage       string              `xml:"errMessage"`
	ItemDetail       *queryItemDetailXML `xml:"invDocCargoVO"`
}

type queryItemDetailXML struct {
	TrackingUrl       string          `xml:"trackingUrl"`
	ReceiverCustName  string          `xml:"receiverCustName"`
	DepartureUnitName string          `xml:"departureUnitName"`
	DeliveryDate      string          `xml:"deliveryDate"`
	DeliveryTime      string          `xml:"deliveryTime"`
	CargoHistory      []cargoEventXML `xml:"invDocCargoDAMVOArray>InvDocCargoDAMVO"`
}

type cargoEventXML struct {
	UnitName   string `xml:"unitName"`
	EventName  string `xml:"eventName"`
	ReasonName string `xml:"reasonName"`
	EventDate  string `xml:"eventDate"`
	EventTime  string `xml:"eventTime"`
	CityName   string `xml:"cityName"`
	TownName   string `xml:"townName"`
}

// QueryShipment sends a query shipment SOAP request and returns the result.
func QueryShipment(client *http.Client, endpoint, username, password, language string, keys []string, keyType int, addHistoricalData bool, onlyTracking bool) (*models.QueryResult, error) {
	soapXML := buildQueryShipmentXML(username, password, language, keys, keyType, addHistoricalData, onlyTracking)

	req, err := http.NewRequest("POST", endpoint, bytes.NewBufferString(soapXML))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "text/xml; charset=utf-8")
	req.Header.Set("SOAPAction", "")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("SOAP fault: HTTP %d - %s", resp.StatusCode, string(body))
	}

	var envelope queryResponseEnvelope
	if err := xml.Unmarshal(body, &envelope); err != nil {
		return nil, fmt.Errorf("failed to parse response XML: %w", err)
	}

	result := &models.QueryResult{
		OutFlag:      envelope.Body.QueryResponse.QueryResult.OutFlag,
		OutResult:    envelope.Body.QueryResponse.QueryResult.OutResult,
		SenderCustId: envelope.Body.QueryResponse.QueryResult.SenderCustId,
		Count:        envelope.Body.QueryResponse.QueryResult.Count,
	}

	for _, d := range envelope.Body.QueryResponse.QueryResult.Details {
		detail := models.QueryDetail{
			CargoKey:         d.CargoKey,
			InvoiceKey:       d.InvoiceKey,
			OperationStatus:  d.OperationStatus,
			OperationMessage: d.OperationMessage,
			ErrCode:          d.ErrCode,
			ErrMessage:       d.ErrMessage,
		}

		if d.ItemDetail != nil {
			itemDetail := &models.QueryItemDetail{
				TrackingUrl:       d.ItemDetail.TrackingUrl,
				ReceiverCustName:  d.ItemDetail.ReceiverCustName,
				DepartureUnitName: d.ItemDetail.DepartureUnitName,
				DeliveryDate:      d.ItemDetail.DeliveryDate,
				DeliveryTime:      d.ItemDetail.DeliveryTime,
			}

			for _, e := range d.ItemDetail.CargoHistory {
				itemDetail.CargoHistory = append(itemDetail.CargoHistory, models.CargoEvent{
					UnitName:   e.UnitName,
					EventName:  e.EventName,
					ReasonName: e.ReasonName,
					EventDate:  e.EventDate,
					EventTime:  e.EventTime,
					CityName:   e.CityName,
					TownName:   e.TownName,
				})
			}

			detail.ItemDetail = itemDetail
		}

		result.Details = append(result.Details, detail)
	}

	return result, nil
}
